aten/src/ATen/native/cuda/Shape.cu

```diff
@@ -0,0 +1,399 @@
+#include <ATen/ATen.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/DeviceGuard.h>
+#include <ATen/Context.h>
+#include <ATen/Utils.h>
+#include <ATen/MemoryOverlap.h>
+#include <ATen/cuda/detail/IndexUtils.cuh>
+#include <ATen/Dispatch.h>
+
+#include <c10/core/TensorImpl.h>
+#include <c10/core/ScalarType.h>
+#include <c10/core/UndefinedTensorImpl.h>
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <c10/util/typeid.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <THC/THC.h>
+
+namespace at {
+namespace native {
+
+#define CAT_ARRAY_BATCH_SIZE 1024
+#define CAT_ARRAY_MAX_INPUT_DIMS 4
+
+namespace {
+
+inline bool getCatGrid(THCState* state, ptrdiff_t nTensors, dim3& grid) {
```

Comment:
why do you need THCState?  We never use this in ATen AFAIK.

Reply:
I didn't know how to replace the memory allocation and deallocation. (`THCudaMalloc`, `THCudaHostAlloc`, `THCudaHostRecord`, `THCudaFree`) `git log -S` turned up `1c0fbd27a10b7036b1db643014e65cae7f6d0266` which still uses `THCudaMalloc`.